### PR TITLE
ENH: Create an experimental export of the new DType API

### PR DESCRIPTION
**Pull request by seberg**
This expposes an experimental DType and ufunc API.
It can be used by including the experimental header AND
setting the `NUMPY_EXPERIMENTAL_DTYPE_API=1` environment variable.

A user must further call:

    if (import_experimental_dtype_api(version) < 0) {
        return NULL;
    }

with a strict version check that will error on any mismatch.  The exposure works using an API table that is fetched through the a python call to `np.core._multiarray_umath. _get_experimental_dtype_api(version)` so that we can always change things.

At this time, I _know_ that some changes will happen, but it feels like
it may be a good idea to make this public anyway, with all the problems
attached...  (e.g. as of now ufunc reductions will just not work, since
that PR is not yet merged.)

---

The exposed code here generally works, which I tested using https://github.com/seberg/experimental_user_dtypes.  To be clear, things will change at least somewhat and some things are likely broken.  But also, most things will stay identical.

As a symbol of things moving, I think I would like to move this forward (it is even usable! :)).  And I think we had some agreement that hiding between `NUMPY_EXPERIMENTAL_DTYPE_API=1` and requiring inclusion of headers with "experimental" in the name should be enough warning? :)

### CODE REVIEW

**Comment by seberg** [comment]
Notes for the review:
1. Changes in `array_method.c` are bug-fixes and simplifications in code that was practically unused and is not used for NumPy proper.
2. There is still `PyArray_DescrNew`, etc. C-API which will just do nonsense for new-style DTypes.  There is a reason to call it experimental :).

**Comment by seberg** [comment]
Could I incentivize someone to have a look?  This will is very much experimental and locked behind checks, so there is no way of breaking all of NumPy.  So the main commodity in reviewing would be some interest in the new API I think ;).

File: numpy/core/src/multiarray/dtypemeta.h [hunk 1]
```diff
@@ -8,13 +8,48 @@
...
+ * the type, or allow looking at the object. A possible use-case:
+ * `np.array(np.array([0]), dtype=np.ndarray)`
+ * Could consider arrays that are not `dtype=ndarray` "scalars".
+ */
+typedef int (is_known_scalar_type_function)(
+        PyArray_DTypeMeta *cls, PyTypeObject *obj);
+
+typedef PyArray_Descr *(default_descr_function)(PyArray_DTypeMeta *cls);
+typedef PyArray_DTypeMeta *(common_dtype_function)(
+        PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtyep2);
```
Comment (rossbar) [inline numpy/core/src/multiarray/dtypemeta.h:25]:
```suggestion
        PyArray_DTypeMeta *dtype1, PyArray_DTypeMeta *dtype2);
```

File: numpy/core/src/multiarray/dtypemeta.h [hunk 2]
```diff
@@ -8,13 +8,48 @@
...
+typedef PyArray_Descr *(common_instance_function)(
+        PyArray_Descr *dtype1, PyArray_Descr *dtyep2);
```
Comment (rossbar) [inline numpy/core/src/multiarray/dtypemeta.h:27]:
```suggestion
        PyArray_Descr *dtype1, PyArray_Descr *dtype2);
```

File: numpy/core/src/multiarray/experimental_public_dtype_api.c [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+    }
+    /*
+     * Lets not trust that the DType is implemented correctly
+     * TODO: Should probably do an exact type-check (at least unless this is
+     *       an abstract DType).
+     */
+    if (!PyArray_DescrCheck(res)) {
+        PyErr_SetString(PyExc_RuntimeError,
+                "Instantiating %S did not return a dtype instance, this is "
+                "invalid (especially without a custom `default_descr()`).");
```
Comment (rossbar) [inline numpy/core/src/multiarray/experimental_public_dtype_api.c:63]: Do you need to add `res` here for the `%S`?

File: numpy/core/src/multiarray/experimental_public_dtype_api.c [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+        if (res < 0) {
+            return -1;
+        }
+    }
+
+    if (NPY_DT_SLOTS(DType)->within_dtype_castingimpl == NULL) {
+        /*
+         * We expect this for now. We should have a default for DType that
+         * only support simple copy (and possibly byte-order when assuming that
+         * they swap the full itemsize).
```
Comment (rossbar) [inline numpy/core/src/multiarray/experimental_public_dtype_api.c:280]:
```suggestion
         * only supports simple copy (and possibly byte-order assuming that
         * they swap the full itemsize).
```

**Review by rossbar — APPROVED** [review]
IMO the earlier the experimental dtype API is made accessible to users, the better! I think the combination of the naming scheme (i.e. "experimental_public_dtype_api") and environment variable is adequate to make sure that anyone who is using the newly exposed functionality is doing so intentionally (and with the understanding that it's subject to change).

All of my inline comments are of the type-edit variety, feel free to modify as you see fit!

File: numpy/core/src/multiarray/experimental_public_dtype_api.c (hunk 3 above)
Comment (seberg) [inline numpy/core/src/multiarray/experimental_public_dtype_api.c:63]:
```suggestion
                "invalid (especially without a custom `default_descr()`).",
                self);
```

**Pull request merged**